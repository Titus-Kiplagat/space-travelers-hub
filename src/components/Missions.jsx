/** @format */

import { Table, Button, Badge } from 'react-bootstrap';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchMissions,
  joinMission,
  leaveMission,
} from '../redux/missionsSlice';

const Missions = () => {
  const { missions, loading, error } = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  if (loading) {
    return <h1 className="container align-middle text-center">Loading...</h1>;
  }
  if (error) {
    return <h1 className="container align-middle text-center">{error}</h1>;
  }
  return (
    <Table striped className="container">
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th colSpan={4}>Status</th>
        </tr>
      </thead>
      <tbody>
        {missions.map(({
          missionId, missionName, description, reserved,
        }) => (
          <tr key={missionId}>
            <td className="bold">{missionName}</td>
            <td>{description}</td>
            <td className="align-middle text-center">
              {reserved ? (
                <Badge className="bold bg-primary" variant="secondary">
                  Active Member
                </Badge>
              ) : (
                <Badge className="bold bg-secondary" variant="secondary">
                  NOT A MEMBER
                </Badge>
              )}
            </td>
            <td className="align-middle text-center">
              {reserved ? (
                <Button
                  size="lg"
                  className="w-100 btn-outline-danger border-2 bold"
                  variant="light"
                  onClick={() => dispatch(leaveMission(missionId))}
                >
                  Leave Mission
                </Button>
              ) : (
                <Button
                  size="lg"
                  className="w-100 btn-outline-secondary border-2 bold"
                  variant="light"
                  onClick={() => dispatch(joinMission(missionId))}
                >
                  Join Mission
                </Button>
              )}
            </td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

export default Missions;
