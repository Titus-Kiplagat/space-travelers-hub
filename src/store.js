import { configureStore } from '@reduxjs/toolkit';
import missionsReducer from './redux/missionsSlice';

const store = configureStore({
  reducer: {
    missions: missionsReducer,
  },
});

export default store;
