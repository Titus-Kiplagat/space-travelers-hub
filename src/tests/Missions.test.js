import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../components/Missions';
import store from '../store';

it('matches snapshot', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
