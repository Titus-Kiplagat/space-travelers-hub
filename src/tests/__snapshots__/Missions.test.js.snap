// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="px-5"
>
  <table
    className="container-fluid table table-striped table-bordered"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th
          colSpan={4}
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
